cnxarchive/oai.py

```diff
@@ -0,0 +1,267 @@
+    new_vars = _verifyMetadataPrefix(request)
+    if 'error' in new_vars.keys():
+        return new_vars
+    where, arguments = _setFromUntil(request)
+    statement = """
+                SELECT revised, uuid
+                FROM latest_modules {}
+                ORDER BY revised;
+                """.format(where)
+    settings = request.registry.settings
+    search_results = _databaseDictResults(statement, arguments, settings)
+    if len(search_results) == 0:
+        return _noRecordsMatchError()
+    new_vars.update({'results': search_results})
+    return new_vars
+
+
+def do_GetRecords(request):
+    new_vars = _verifyMetadataPrefix(request)
+    if 'error' in new_vars.keys():
+        return new_vars
+    identifier = str(request.GET.get('identifier'))
+    idDoesNotExist = {'code': 'idDoesNotExist',
+                      'message': "id does not exist {}".
+                                 format(identifier)}
+    pattern = ("oai:" + request.host + ":[0-9a-z]{8}-[0-9a-z]{4}-"
+               "[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{12}$")
+    if not compile(pattern).match(identifier):
+        return {'error': idDoesNotExist}
+    uuid = identifier.split(':')[-1]
+    statement = "SELECT" + _oaiGetBaseStatement()
+    statement += "FROM latest_modules as lm  WHERE uuid=(%s);"
+    settings = request.registry.settings
+    results = _databaseDictResults(statement, (uuid,), settings)
+    if len(results) == 0:
+        return {'error': idDoesNotExist}
+    new_vars.update(_formatOaiResults(results, request))
+    return new_vars
+
+
+def do_ListRecords(request):
+    new_vars = _verifyMetadataPrefix(request)
+    if 'error' in new_vars.keys():
+        return new_vars
+    where, arguments = _setFromUntil(request)
+    statement = "SELECT" + _oaiGetBaseStatement()
+    statement += "FROM latest_modules as lm {} ORDER BY revised;".format(where)
+    settings = request.registry.settings
+    results = _databaseDictResults(statement, arguments, settings)
+    new_vars.update(_formatOaiResults(results, request))
+    return new_vars
+
+
+# ################## #
+#      OAI VIEW      #
+# ################## #
+
+
+@view_config(route_name='oai', request_method='GET',
+             renderer='templates/oai_verbs.xml')
+def oai(request):
+    request.response.headers = {'Content-Type': 'text/xml; charset=UTF-8'}
+
+    return_vars = {}
+    return_vars['dateTime'] = html_rss_date(datetime.now().time())
+    return_vars['baseURL'] = request.path_url
+    return_vars['query_request'] = []
+    for var in request.GET:
+        return_vars['query_request'].append({'name': var,
+                                             'val': request.GET[var]})
+
+    verb = request.GET.get('verb')
+    parse_results = _parseOaiArguments(verb, request)
+    return_vars['verb'] = verb
+    if parse_results != "No errors":
+        return_vars['error'] = parse_results
+        return return_vars
+
+    oaiQuery = {'Identify': do_Identify(request),
```

Comment:
This dict would stay in oai.py, and be imported from there by views.py for use here.